### DISCUSSION TITLE
Some potential `bindToDevice` issues

### DISCUSSION THREAD

**Discussion by dyhkwong** [v2fly/v2ray-core#2878] [General]
There may be some issues for `bindToDevice` but I did no further investigation.
- According to Microsoft documents and other proxy software, interface index should be a `DWORD` (uint32) in **network byte order** for `IP_UNICAST_IF`, and in host byte order for `IPV6_UNICAST_IF`.
https://github.com/v2fly/v2ray-core/blob/458db3954f76bbf7ce621a24151d9d5469d8b6fb/transport/internet/sockopt_windows.go#L42-L53
- On Darwin and Windows, it tries to set **both** `IP_(UNICAST|BOUND)_IF` and `IPV6_(UNICAST|BOUND)_IF` for a single socket without ignoring expected errors. Normally only one of them should be set.
- On Windows, only `applyOutboundSocketOptions` calls `bindToDevice`. But on Darwin and Linux both `applyOutboundSocketOptions` and `applyInboundSocketOptions` call `bindToDevice`.

**Comment by grqz** [comment]
The binary patch below seems to fix the windows IP_UNICAST_IF issue.

<details>

<summary>IPS patch (xxd dump) for v2ray.exe unpacked from v2ray-windows-64.zip, v5.53.0</summary>

```
00:50 41 54 43 48    ; PATCH

05:4c 3e 32
08:00 14             ; size = 20

0a:48 8d 7c 24 30    ; lea rdi, [rsp+0x30]
0f:6a 1f             ; push 0x1f
11:59                ; pop rcx
12:8b 37             ; mov esi, dword [rdi]
14:0f ce             ; bswap esi
16:89 37             ; mov dword [rdi], esi
18:90                ; nop
19:e8 8c 04 00 00    ; call stub

1e:4c 42 d2          ; stub:
21:00 08             ; size = 8

23:6a 04             ; push 0x4
25:5e                ; pop rsi
26:e9 86 52 ff ff    ; jmp SetsockoptInt

2b:45 4f 46          ; END
```

</details>
